tools/runtime-condenser/src/main.rs

```diff
@@ -0,0 +1,171 @@
+use std::fs::File;
+use std::io::prelude::*;
+use std::io::BufReader;
+use std::collections::hash_map::{HashMap, Entry};
+
+const INPUT_FILE: &str = "input.txt";
+const OUTPUT_FILE: &str = "output.txt";
+
+type Runtimes = HashMap<String, RuntimeData>;
+
+struct RuntimeData {
+	pub details: String,
+	pub counter: usize,
+	pub kind: RuntimeKind
+}
+
+enum RuntimeKind {
+	RuntimeError,
+	InfiniteLoop,
+	RecursionLimit
+}
+
+enum State {
+	Runtime(String),
+	Skip(usize),
+	Scanning
+}
+
+impl std::fmt::Display for RuntimeData {
+	fn fmt(&self, f: &mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
+		write!(f, "{}", self.details)?;
+		writeln!(f, "Occurred {} times.", self.counter)?;
+		Ok(())
+	}
+}
+
+fn insert_runtime(runtimes: &mut Runtimes, identifier: String, kind: RuntimeKind) {
+	match runtimes.entry(identifier) {
+		Entry::Vacant(entry) => {
+			entry.insert(RuntimeData {
+				details: String::new(),
+				counter: 1,
+				kind: kind
+			});
+		}
+		Entry::Occupied(mut entry) => {
+			let data = entry.get_mut();
+			data.counter += 1;
+		}
+	};
+}
+
+fn line_kind(line: &str) -> LineKind {
+	if line.find("Infinite loop suspected--switching proc to background.").is_some() {
+		return LineKind::InfiniteLoopHeader;
+	}
+	if line.starts_with("proc name") {
+		return LineKind::InfiniteLoop;
+	}
+	if line.find("Runtime in ").is_some() {
+		return LineKind::Runtime;
+	}
+	if line.starts_with("  ") {
+		return LineKind::Details;
+	}
+	LineKind::Junk
+}
+
+enum LineKind {
+	InfiniteLoopHeader,
+	InfiniteLoop,
+	Runtime,
+	Details,
+	Junk
+}
+
+fn parse_from_file(path: &str) -> Runtimes {
+	let input_file = File::open(path).expect("Error opening file.");
+	let reader = BufReader::new(input_file);
+	let lines = reader.lines().map(std::result::Result::unwrap);
+	
+	let mut runtimes = Runtimes::new();
+	let mut current_state = State::Scanning;
+
+	for line in lines {
+		if let State::Skip(mut count) = current_state {
+			count = count - 1;
+			if count < 1 {
+				current_state = State::Scanning
+			}
+			continue;
+		}
+		match line_kind(&line) {
+			LineKind::InfiniteLoopHeader => {
+				current_state = State::Skip(1);
+			}
+			LineKind::InfiniteLoop => {
+				if let State::Scanning = current_state {
+					current_state = State::Runtime(line.clone());
+					insert_runtime(&mut runtimes, line, RuntimeKind::InfiniteLoop);
+				}
+			}
+			LineKind::Details => {
+				if let State::Runtime(ref ident) = current_state {
+					let data = runtimes.get_mut(ident).unwrap();
+					if data.counter != 1 {
+						continue;
+					}
+					data.details.push_str(&line);
+					data.details.push('\n');
+				}
+			}
+			LineKind::Runtime => {
+				let runtime_identifier = line[22..].to_owned();
+				current_state = State::Runtime(runtime_identifier.clone());
+				insert_runtime(&mut runtimes, runtime_identifier, RuntimeKind::RuntimeError);
```

Comment:
You're cloning the identifier *twice* before checking whether that was even necessary (inside `insert_runtime`). It's all wasted if it only needed to increase the counter.